internal/internal_workflow.go

```diff
@@ -646,11 +648,20 @@ func getState(ctx Context) *coroutineState {
 	}
 	state := s.(*coroutineState)
 	if !state.dispatcher.IsExecuting() {
-		panic(panicIllegalAccessCoroutinueState)
+		panic(panicIllegalAccessCoroutineState)
 	}
 	return state
 }
 
+func assertNotInReadOnlyState(ctx Context) {
```

Comment:
Should this be invoked just before command add? Or is there not a reasonable place to do that? (sorry if I'm commenting too early)

Reply:
Not really a reasonable place, there are also edge cases when functions that can't be called in read only context only sometimes generate commands, or functions that don't generate commands but still shouldn't be called in read only context. I think it is easier to implement, allows us the fail faster, and reads clearer ,because it is more consistent, if every function you shouldn't call has its own assert.

Reply:
Right, I can understand adding additional checks for things that don't send commands.

> if every function you shouldn't call has its own assert.

If you don't think it'll be too crazy or hard to maintain, works for me. Put it as close to the outer call boundaries as we can (i.e. right as first line of exported thing, not behind interceptors or after other validation or whatever).